piet-gpu-hal/src/hub.rs

```diff
@@ -692,6 +722,32 @@ impl Drop for SubmittedCmdBuf {
     }
 }
 
+impl<'a> ComputePass<'a> {
+    /// Dispatch a compute shader.
+    ///
+    /// Request a compute shader to be run, using the pipeline to specify the
+    /// code, and the descriptor set to address the resources read and written.
+    ///
+    /// Both the workgroup count (number of workgroups) and the workgroup size
+    /// (number of threads in a workgroup) must be specified here, though not
+    /// all back-ends require the latter info.
+    pub unsafe fn dispatch(
+        &mut self,
+        pipeline: &Pipeline,
+        descriptor_set: &DescriptorSet,
+        workgroup_count: (u32, u32, u32),
+        workgroup_size: (u32, u32, u32),
+    ) {
+        self.cmd_buf
+            .cmd_buf()
+            .dispatch(pipeline, descriptor_set, workgroup_count, workgroup_size);
+    }
+
+    pub unsafe fn end(&mut self) {
```

Comment:
Should this consume self? It seems like the right choice semantically but there might be ergonomic issues of which I'm unaware.

Reply:
Good question. The underlying issue is that you can't easily have an encoder embedded in some other struct, otherwise you get errors like "cannot move out of." It's possible to address those in turn by writing `Option<Encoder>` and doing a `take`, but that's clunky. I ran into similar problems in resource (image, buffer) destructors but now that I think about it, we might not have the same problem. It's likely that `end` will be called in the same scope as encoder creation. I'll make the change and we can revert it if it turns out to cause problems.

I'll also highlight some ergonomics issues. For a lot of the "stage" stuff in piet-gpu, we'll need to pass in a compute encoder, perhaps instead of a command buffer, perhaps in addition to it. Things get more complicated if any of those stages need to do blit or clear operations; I haven't carefully audited the code. If so, then there will need to be more than one compute encoder. It should still be doable to have results from multiple encoders into a single query pool, but I still would like to make the change of returning timestamps directly rather than subtracting off the first one.

Reply:
> I'll also highlight some ergonomics issues. For a lot of the "stage" stuff in piet-gpu, we'll need to pass in a compute encoder, perhaps instead of a command buffer, perhaps in addition to it. Things get more complicated if any of those stages need to do blit or clear operations; I haven't carefully audited the code.

Followup: all uses were clean with the stages requiring only dispatches and memory barriers so the conversion to ComputePass (with consuming self) was fairly trivial.